### PR TITLE
Avoid `get_notes` call in `CouponPageMoved`

### LINKED ISSUE #7986: Remove expensive `get_notes` call in the `CouponPageMoved::possibly_add_note` function
### Describe

As mentioned in #7968, the `CouponPageMoved::possibly_add_note` makes use of the `get_notes` function that can be expensive as it contains several filters for 3PD's and is also called on `admin_init`.
Given this call is very specific to the `CouponPageMoved` note and checking if it is deleted or not, we should do this check ourselves using $wpdb.

### Acceptance criteria

- Make sure the `CouponPageMoved::possibly_add_note` doesn't make use of the `get_notes` function.

### PR DESCRIPTION
**Pull request by jacob-sewell**
Fixes #7986 

`CouponPageMoved::possibly_add_note()` was potentially calling `get_notes()` from the data store twice, which can get expensive. I have changed both the called methods (`has_dismissed_note()` and `has_unactioned_note()`) to directly call `$wpdb` with simpler, less expensive queries.

### Detailed test instructions:

The relevant method runs as part of the [admin_init](https://developer.wordpress.org/reference/hooks/admin_init/) hook, so among other things it happens any time you load wp-admin. If coupons are enabled, legacy coupon menu is enabled, and there isn't already an unactioned or soft-deleted `wc-admin-coupon-page-moved` note in the database, one should be created.

### CODE REVIEW

**Comment by jacob-sewell** [comment]
@louwie17 , correct me if I'm misunderstanding, but it seems like the biggest issue is the fact that `get_notes()` hits the `'woocommerce_note_where_clauses'` filter via `get_notes_where_clauses()`, yeah? This should avoid that.

I don't love the near-duplication of the `$wpdb->get_var()` calls here, so I may take a refactor pass, but I want to sanity-check this approach first.

File: src/Notes/CouponPageMoved.php [hunk 1]
```diff
@@ -94,15 +94,16 @@ public static function get_note() {
...
 	protected static function has_unactioned_note() {
-		$notes = self::get_data_store()->get_notes(
-			[
-				'name'       => [ self::NOTE_NAME ],
-				'status'     => [ 'unactioned' ],
-				'is_deleted' => false,
-			]
+		global $wpdb;
+
+		$unactioned_note_count = $wpdb->get_var(
```
Comment (joshuatf) [inline src/Notes/CouponPageMoved.php:99]: Could we use `Notes::get_note()` to retrieve these notes? 

In general, we should try our best to limit custom queries like this where possible.

File: src/Notes/CouponPageMoved.php [hunk 2]
```diff
@@ -111,14 +112,16 @@ protected static function has_unactioned_note() {
...
 	 */
 	protected static function has_dismissed_note() {
-		$notes = self::get_data_store()->get_notes(
-			[
-				'name'       => [ self::NOTE_NAME ],
-				'is_deleted' => true,
-			]
+		global $wpdb;
+
+		$dismissed_note_count = $wpdb->get_var(
```
Comment (joshuatf) [inline src/Notes/CouponPageMoved.php:117]: Same comment here.

File: src/Notes/CouponPageMoved.php (hunk 1 above)
Comment (jacob-sewell) [reply src/Notes/CouponPageMoved.php:99]: We could use `Notes::get_note()` if we knew the `$note_id`, but we don't.

**Comment by jacob-sewell** [comment]
I went with using `$wpdb` directly because that's what was suggested in the issue, but I agree it's a little smelly in an environment where we have a Data Store. But if I generalize directly looking up database rows without incurring the expense of the `woocommerce_note_where_clauses` hook (e.g. `get_notes_no_filters()`) don't I introduce lots of potential regressions with third party code?

File: src/Notes/CouponPageMoved.php (hunk 1 above)
Comment (joshuatf) [reply src/Notes/CouponPageMoved.php:99]: In that case you can use:

```
		$data_store = Notes::load_data_store();
		$note_ids   = $data_store->get_notes_with_name( self::NOTE_NAME );
		if ( count( $note_ids ) > 0 ) {
				$note = Notes::get_note( $note_ids[0] )
		}
```

Also, not a necessity in this PR, but I would not at all be opposed to having a helper function do this.  We've used this pattern quite a bit of getting note IDs and looping over them; seems like a simply `get_note_by_name()` would not be a bad thing.

**Review by joshuatf — COMMENTED** [review]
>But if I generalize directly looking up database rows without incurring the expense of the woocommerce_note_where_clauses hook (e.g. get_notes_no_filters()) don't I introduce lots of potential regressions with third party code?

Great question.  The original issue does sound more to me like one that is a bug in third party plugins and not necessarily an issue with searching for notes.

That said, an API method to fetch a note by name (or using the existing pattern of get notes by name and then grabbing the note) would be preferred to manual querying if we choose to change this here since that could be brittle with changes to the notes schema.

File: src/Notes/Notes.php [hunk 3]
```diff
@@ -82,6 +82,26 @@ public static function get_note( $note_id ) {
...
+	/**
+	 * Get admin note using its name.
+	 *
+	 * This is a shortcut for the common pattern of looking up note ids by name and then passing the first id to get_note().
+	 * It will behave unpredictably when more than one note with the given name exists.
+	 *
+	 * @param string $note_name Note name.
+	 * @return Note|bool
+	 **/
+	public static function get_note_by_name( $note_name ) {
```
Comment (joshuatf) [inline src/Notes/Notes.php:94]: 🎉

**Review by joshuatf — APPROVED** [review]
Nice work on this, Jacob!  Testing well and love the code improvements.  🚢

**Pull request merged**